opentelemetry-exporter-geneva/geneva-uploader/src/payload_encoder/lz4_chunked_compression.rs

```diff
@@ -0,0 +1,177 @@
+///  | 04 |....    | 03 |...    | 02 |..   |
+///  +----+--------+----+-------+----+-----+
+///   ^len1 data1   ^len2 data2   ^len3 data3
+/// ```
+///
+/// # Notes
+/// - This chunked format is **not required by LZ4** itself, but is a common convention to allow boundary detection.
+/// - Decompression requires reading 4 bytes, then decompressing the next `len` bytes, then repeating.
+/// - This function is allocation-efficient: only two buffers are allocated (one for output, one for temp).
+/// # TODO
+/// - Investigate if true in-place compression is possible when the input is no longer needed.
+///   This might let us use the input buffer as the temporary compression buffer, potentially reducing heap allocations further,
+///   but care must be taken: LZ4 does not natively support in-place compression, and the compressed size may be larger than the input.
+///   If single-buffer "in-place" compression is possible (e.g., with unsafe or buffer aliasing), document or implement it here.
+/// - Consider passing output buffer as mutable slice to avoid reallocation, and provide another method to return the max size of the output buffer.
+#[allow(dead_code)]
+pub(crate) fn lz4_chunked_compression(
+    input: &[u8],
+) -> Result<Vec<u8>, lz4_flex::block::CompressError> {
+    const CHUNK_SIZE: usize = 64 * 1024;
+    let max_chunk_compressed = get_maximum_output_size(CHUNK_SIZE);
+
+    // Pre-allocate an output buffer large enough for the worst-case total output size:
+    // Each chunk may require up to get_maximum_output_size(CHUNK_SIZE) bytes for compressed data,
+    // plus 4 bytes for the length header per chunk.
+    let mut output =
+        Vec::with_capacity(input.len().div_ceil(CHUNK_SIZE) * (4 + max_chunk_compressed));
+
+    let mut offset = 0;
+    // Process the input in 64 KiB chunks.
+    while offset < input.len() {
+        // Determine the end index for the current chunk.
+        let end = usize::min(offset + CHUNK_SIZE, input.len());
+        // Get the current chunk from input.
+        let chunk = &input[offset..end];
+
+        // Reserve space for the 4-byte header
+        let header_offset = output.len();
+        output.extend_from_slice(&[0u8; 4]); // Placeholder for length
+
+        // Reserve worst-case space for compressed data
+        let data_offset = output.len();
+        output.resize(data_offset + max_chunk_compressed, 0);
+
+        // Compress directly into the reserved slice
+        let compressed_size = compress_into(
+            chunk,
+            &mut output[data_offset..data_offset + max_chunk_compressed],
+        )?;
+
+        // Write the actual compressed length as little-endian u32
+        let compressed_size_le = (compressed_size as u32).to_le_bytes();
+        output[header_offset..header_offset + 4].copy_from_slice(&compressed_size_le);
+        // Truncate output to actual size (header + compressed)
+        output.truncate(data_offset + compressed_size);
```

Comment:
This can be optimized further without needing to resize and truncate during each iteration.

Reply:
Add a TODO for this if you want to address this later.

Reply:
Added TODO for now, will optimize it in subsequent PR for encoder.